### PR TITLE
Charge Card Transaction - Workflow changes

### PR DESCRIPTION
**Pull request by ashwin1111**

### CODE REVIEW

File: apps/sage_intacct/migrations/0008_auto_20210408_0812.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+    dependencies = [
+        ('sage_intacct', '0007_auto_20210308_0759'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='chargecardtransaction',
+            name='reference_no',
+            field=models.CharField(default='dummy_reference_no', help_text='Sage Intacct Reference number to transaction', max_length=255),
```
Comment (ashwin1111) [inline apps/sage_intacct/migrations/0008_auto_20210408_0812.py:16]: defaulting to a dummy value, fyi

File: apps/sage_intacct/migrations/0008_auto_20210408_0812.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+            preserve_default=False,
+        ),
+        migrations.AddField(
+            model_name='chargecardtransaction',
+            name='vendor_id',
+            field=models.CharField(default='dummy_vendor_id', help_text='Sage Intacct Vendor ID', max_length=255),
```
Comment (ashwin1111) [inline apps/sage_intacct/migrations/0008_auto_20210408_0812.py:22]: defaulting to a dummy value, fyi

File: apps/sage_intacct/models.py [hunk 3]
```diff
@@ -500,12 +519,12 @@ def create_charge_card_transaction(expense_group: ExpenseGroup):
                 expense_group=expense_group,
                 defaults={
                     'charge_card_id': charge_card_id,
+                    'vendor_id': vendor,
                     'description': description,
-                    'memo': 'Reimbursable expenses by {0}'.format(description.get('employee_email')) if
-                    expense_group.fund_source == 'PERSONAL' else
-                    'Credit card expenses by {0}'.format(description.get('employee_email')),
+                    'memo': 'Credit card expenses by {0}'.format(description.get('employee_email')),
```
Comment (ashwin1111) [inline apps/sage_intacct/models.py:524]: CCT can only be created for CCC expenses

File: apps/sage_intacct/models.py [hunk 4]
```diff
@@ -500,12 +519,12 @@ def create_charge_card_transaction(expense_group: ExpenseGroup):
...
+                    'reference_no': expense.expense_number,
                     'currency': expense.currency,
-                    'transaction_date': get_transaction_date(expense_group)
+                    'transaction_date': expense_group.description['spent_at']
```
Comment (ashwin1111) [inline apps/sage_intacct/models.py:527]: CCT can only be created for CCC expenses

File: apps/sage_intacct/tasks.py [hunk 5]
```diff
@@ -301,21 +323,6 @@ def __validate_expense_group(expense_group: ExpenseGroup, general_settings: Work
             'message': 'General mappings not found'
         })
 
-    general_settings: WorkspaceGeneralSettings = WorkspaceGeneralSettings.objects.get(
-        workspace_id=expense_group.workspace_id)
-
-    if general_settings.corporate_credit_card_expenses_object:
```
Comment (ashwin1111) [inline apps/sage_intacct/tasks.py:307]: removed redundant method

File: apps/sage_intacct/tasks.py [hunk 6]
```diff
@@ -301,21 +323,6 @@ def __validate_expense_group(expense_group: ExpenseGroup, general_settings: Work
             'message': 'General mappings not found'
         })
 
-    general_settings: WorkspaceGeneralSettings = WorkspaceGeneralSettings.objects.get(
```
Comment (ashwin1111) [inline apps/sage_intacct/tasks.py:304]: we already get this in fn args

File: apps/sage_intacct/utils.py [hunk 7]
```diff
@@ -257,6 +257,41 @@ def sync_employees(self):
...
+
+    def get_or_create_vendor(self, vendor_name: str, email: str = None, create: bool = False):
+        """
+        Call Sage Intacct api to get or create vendor
+        :param vendor_name: Name of the vendor
+        :param email: Email of the vendor
+        :param create: False to just Get and True to Get or Create if not exists
+        :return: Vendor
+        """
+        vendor = self.connection.vendors.get(field='NAME', value=vendor_name.replace("'", "\\'"))
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:281]: intacct accepts `#`

File: scripts/007-update-expense-group-settings.sql [hunk 8]
```diff
@@ -0,0 +1,5 @@
+rollback;
+begin;
+
+update expense_group_settings
```
Comment (ashwin1111) [inline scripts/007-update-expense-group-settings.sql:4]: enable this for all orgs

File: apps/sage_intacct/utils.py [hunk 9]
```diff
@@ -257,7 +257,44 @@ def sync_employees(self):
...
+        if 'vendor' in vendor:
+            vendor = vendor['vendor'][0] if int(vendor['@totalcount']) > 1 else vendor['vendor']
+        else:
+            vendor = None
+
+        if not vendor:
+            if create:
+                created_vendor = self.post_vendor(vendor_name, email)
+                return self.create_vendor_destionation_attribute(created_vendor['VENDORID'],
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:290]: Vendor ID and name would be the same

File: apps/sage_intacct/models.py [hunk 10]
```diff
@@ -458,8 +458,10 @@ class ChargeCardTransaction(models.Model):
     id = models.AutoField(primary_key=True)
     expense_group = models.OneToOneField(ExpenseGroup, on_delete=models.PROTECT, help_text='Expense group reference')
     charge_card_id = models.CharField(max_length=255, help_text='Sage Intacct Charge Card ID')
+    vendor_id = models.CharField(max_length=255, help_text='Sage Intacct Vendor ID')
     description = models.TextField(help_text='Sage Intacct Charge Card Transaction Description')
     memo = models.TextField(help_text='Sage Intacct referenceno', null=True)
+    reference_no = models.CharField(max_length=255, help_text='Sage Intacct Reference number to transaction')
```
Comment (Shwetabhk) [inline apps/sage_intacct/models.py:464]: can we have duplicate reference numbers in Sage Intacct?

File: apps/sage_intacct/models.py [hunk 11]
```diff
@@ -480,6 +482,23 @@ def create_charge_card_transaction(expense_group: ExpenseGroup):
         expense = expense_group.expenses.first()
 
         if expense_group.fund_source == 'CCC':
```
Comment (Shwetabhk) [inline apps/sage_intacct/models.py:484]: why do we need this check?

Comment (Shwetabhk) [reply apps/sage_intacct/models.py:484]:
```suggestion
        
```

File: apps/sage_intacct/models.py (hunk 4 above)
Comment (Shwetabhk) [reply apps/sage_intacct/models.py:527]: Lets not change this, if spent at is null current date should be exported
```suggestion
                    'transaction_date': get_transaction_date(expense_group)
```

File: apps/sage_intacct/models.py (hunk 10 above)
Comment (ashwin1111) [reply apps/sage_intacct/models.py:464]: yes

File: apps/sage_intacct/models.py (hunk 11 above)
Comment (ashwin1111) [reply apps/sage_intacct/models.py:484]: day 0, this is not required, will remove it off.

File: apps/sage_intacct/utils.py (hunk 9 above)
Comment (Shwetabhk) [reply apps/sage_intacct/utils.py:290]: Didn't get this

Comment (Shwetabhk) [reply apps/sage_intacct/utils.py:290]: Can't we have duplicate Vendor IDs in Intacct? If yes then this wont work no?

Comment (ashwin1111) [reply apps/sage_intacct/utils.py:290]: discussed offline;

**Review by Sravanksk — APPROVED** [review]

**Pull request merged**